### DISCUSSION TITLE
There is a problem with the display of Chinese fonts

### DISCUSSION THREAD

**Discussion by smoosex** [ghostty-org/ghostty#3499] [Q&A] [answered]
this is in ghostty:
![Image](https://github.com/user-attachments/assets/ceed5ba4-ca4d-4ef0-98c5-965a6d470fe9)

this is in kitty:
<img width="289" alt="Image" src="https://github.com/user-attachments/assets/4dc1de03-9a37-425f-af54-8f3dde2d94e5" />

in ghostty, Some Chinese fonts are bold, some are light，I used same font family in ghostty and kitty.

**Comment by KiaraGrouwstra** [comment]
this may not help much, but for what it's worth, it looks like the characters ghostty shows in bold are present in Japanese as well, while the light ones seem to be characters specific to simplified Chinese.

**Comment by pluiedev** [comment]
What font is this? I use [LXGW NeoXiHei](https://github.com/lxgw/LxgwNeoXiHei) and I couldn't reproduce. (My guess would be that font synthesis is broken?)

![Image](https://github.com/user-attachments/assets/6a86dc34-b8c4-433e-929d-0cc8a410c08c)

**Reply by smoosex** [reply]
Previously I used JetBrainsMono.
The following config can make Chinese show Normally.

font-family = "Maple Mono NF"
font-codepoint-map = "U+4E00-U+9FFF,U+3400-U+4DBF = LXGW WenKai"

**Comment by mitchellh** [comment]
This looks like an issue of choosing incorrect fallbacks. Can you use `ghostty +show-face` to print this string and tell me what fonts its choosing?

**Reply by smoosex** [reply]
Previously I used JetBrainsMono.
The following config can make Chinese show Normally.

font-family = "Maple Mono NF"
font-codepoint-map = "U+4E00-U+9FFF,U+3400-U+4DBF = LXGW WenKai"

**Reply by yaoyhu** [reply]
> This looks like an issue of choosing incorrect fallbacks. Can you use `ghostty +show-face` to print this string and tell me what fonts its choosing?

I encountered the same problem. The font shown in the images from this discussion is Apple’s default font designed for Chinese: `PingFang SC`. ~~It seems that `LXGW`, an open-source Chinese font, works well~~ (punctuation has the same issue).

Also, I have a question regarding the `❯` symbol. It renders differently in Ghostty, as shown in the image below, while it appears normal in the VS Code terminal (or any other app, for that matter).

```shell
# Hi world
❯ ghostty +show-face --string="Hi 世界"
U+48 « H » found in face “Fira Code”.
U+69 « i » found in face “Fira Code”.
U+20 «   » found in face “Fira Code”.
U+4E16 « 世 » found in face “PingFang SC”.
U+754C « 界 » found in face “PingFang SC”.
```

<img width="436" alt="image" src="https://github.com/user-attachments/assets/3f1665f9-f279-41a7-bd5d-ab84fb136167" />

**Reply by pluiedev** [reply]
It's most likely that the prompt powerline symbol was drawn with built-in vector graphics instead of using any font you've configured

**Reply by yaoyhu** [reply]
> It's most likely that the prompt powerline symbol was drawn with built-in vector graphics instead of using any font you've configured

It works fine in every other app, but it behaves strangely in Ghostty.

**Reply by acehinnnqru** [reply]
This problem still exists and I don't know how to determine what's the cause.

It would change to a random font after running for a duration (didn't change anything just using).

**Reply by yaoyhu** [reply]
> This problem still exists and I don't know how to determine what's the cause.
> 
> It would change to a random font after running for a duration (didn't change anything just using).

Try [version 1.0.1](https://ghostty.org/docs/install/release-notes/1-0-1)—it works with PingFang SC on my machine.

Sadly, the `❯` symbol remains unchanged.

<img width="591" alt="image" src="https://github.com/user-attachments/assets/0e07d48f-cfaa-42f2-bca4-8053899b2c73" />

**Reply by acehinnnqru** [reply]
> > This problem still exists and I don't know how to determine what's the cause.
> 
> > 
> 
> > It would change to a random font after running for a duration (didn't change anything just using).
> 
> 
> 
> Try [version 1.0.1](https://ghostty.org/docs/install/release-notes/1-0-1)—it works with PingFang SC on my machine.
> 
> 
> 
> Sadly, the `❯` symbol remains unchanged.
> 
> 
> 
> <img width="591" alt="image" src="https://github.com/user-attachments/assets/0e07d48f-cfaa-42f2-bca4-8053899b2c73" />
> 
> 

No, what I met is that the font would change during running but not it can't find the font. It works correctly for a duration and suddenly change to another font to render but the ghostty +show-face still output the correct font.

Restart the ghostty can fixed it but after a duration it would change to render another font...

That's saying that the rendered shape doesn't match the found font face.

**Reply by yaoyhu** [reply]
> It works correctly for a duration and suddenly change to another font to render but the ghostty +show-face still output the correct font.

@acehinnnqru That’s interesting. You might want to provide a screenshot or a video to better illustrate the issue you’re experiencing.

**Reply by acehinnnqru** [reply]
> @acehinnnqru That’s interesting. You might want to provide a screenshot or a video to better illustrate the issue you’re experiencing.

@yaoyhu 

I don't how to trigger to change when running. But I have some screenshots previous.

The correct render behavior: The rendered face is `PingFang SC`.
<img width="559" alt="image" src="https://github.com/user-attachments/assets/1a5f5bd6-72c0-4c52-8e60-b75ffacee101" />

When it change, the rendered change but the `show-face` still says it's `PingFang SC`:
![image](https://github.com/user-attachments/assets/e86b6763-6a5c-48a6-983b-5b623d5b7402)

**Reply by acehinnnqru** [reply]
it changes to another font again:
<img width="403" alt="image" src="https://github.com/user-attachments/assets/886986b5-781f-4992-84d4-3de916bd7de6" />

**Reply by acehinnnqru** [reply]
This is quite annoying. I can't manually reproduce it but I almost met it everyday... I have no idea how to trigger it or how to debug it

**Comment by acehinnnqru** [comment]
I found this issue randomly occured but not always. I didn't set font family, it worked great. But today I found sometimes it would use the wrong fallback. I can't reproduce it.

**Reply by acehinnnqru** [reply]
It fallbacks to another font again, I didn't do anything:
<img width="264" alt="image" src="https://github.com/user-attachments/assets/6d77a4cc-d038-477a-9c1e-50c4045ff716" />

how can I see what had happened? @mitchellh

**Reply by acehinnnqru** [reply]
I was coding in neovim for several hours, and it suddenly changed to another font. It behaves some weird.

**Reply by pluiedev** [reply]
> how can I see what had happened?

Try using the `+show-face` action and see what fonts it's using for each character

**Reply by acehinnnqru** [reply]
> > how can I see what had happened?
> 
> Try using the `+show-face` action and see what fonts it's using for each character

I have tried this and it shows the correct font family. My issue is that it changes when using.

**Reply by zwyyy456** [reply]
> I found this issue randomly occured but not always. I didn't set font family, it worked great. But today I found sometimes it would use the wrong fallback. I can't reproduce it.

I have the same problem, too.

**Reply by acehinnnqru** [reply]
I still meet this issue randomly, I can't reproduce it manually...

**Comment by acehinnnqru** [comment]
This should be a bug:
<img width="414" alt="image" src="https://github.com/user-attachments/assets/40f893a5-b652-4792-abf4-9c96bdc0004f" />

The drawn font isn't "PingFang SC".

**Comment by ShikChen** [comment]
I encountered the same issue with the `PingFang TC` font (the default macOS font for Traditional Chinese). It isn't always reproducible, but the following steps resolved it for me this time:

1. Comment out the `font-family = SauceCodePro Nerd Font Mono` line in the ghosty config.
2. Reload using `Cmd+Shift+,`. The font then falls back to JetBrains Mono, and the bold rendering issue disappears.
3. Restore the original `font-family` line. The Chinese characters continue to display correctly afterward.

**Reply by nyngwang** [reply]
What TC characters did you use to trigger those bold chars? I'm trying to reproduce it with `PingFang TC`, but haven't made it. I tested "字體測試" and my random notes, but those chars didn't go wrong. On the other hand, I needed to use very large `font-size` to read those TC characters clearly because `PingFang TC` made them very small on `font-size = 20`. It would be great if there is some example document that can trigger the bug more frequently. (anyway, just trying to help~)

<img src="https://github.com/user-attachments/assets/cb94faaf-c925-4c55-bd78-cff95838ee7c" width="700">

<details closed>
<summary>related config on testing:</summary>

```text
font-size = 20
font-family = "PingFang TC"
font-style = default
font-style-bold = default
font-style-italic = default
font-style-bold-italic = default
font-synthetic-style = no-bold,italic,no-bold-italic
font-feature =
font-variation =
font-variation-bold =
font-variation-italic =
font-variation-bold-italic =
font-codepoint-map =
font-thicken = true
font-thicken-strength = 0
```
</details>

**Reply by shilohlee98** [reply]
I’ve encountered the same issue, just like @ShikChen — I’m using macOS with PingFang TC (default font is JetBrainsMono Nerd Font).

I’ve also found a way to 100% reproduce the problem. The workaround provided by @ShikChen does solve the bold rendering issue.

Interestingly, I discovered that whether I comment out or uncomment the font-family field and then reload the config, it always fixes the problem.

The key point is: if I close Ghostty using Command+Q, the issue will reappear next time I launch it.

My guess is that there might be a problem when Ghostty initially loads the config — but once you modify font-family and reload, it correctly applies the fallback font?

[reproduce video](https://youtu.be/b0fmjN-UBpQ)

**Reply by jiacai2050** [reply]
This bug also bite me as well, will try to find where the root cause is.

**Comment by hiql** [comment]
same issue, and reproduced when the laptop wakes up every time

**Comment by acehinnnqru** [comment]
Set the font of chinese you like as the last fallback font

**Comment by deeengan** [comment]
i seem to also have this issue on macos

```
/Applications/Ghostty.app/Contents/MacOS/ghostty +show-face --string="h歹"
U+68 « h » found in face “MonaspiceRn Nerd Font Mono”.
U+6B79 « 歹 » found in face “PingFang TC”.
```

but 

```
/Applications/Ghostty.app/Contents/MacOS/ghostty +list-fonts
```

and PingFang TC is not listed

if i try

      font-family = "MonaspiceRn Nerd Font Mono", "Lantinghei TC"

it resets all fonts to JetBrainsMono and my Chinese font is stuck on LXGW WenKai Mono TC, which i am not sure how i set or changed

my rig only changes and displays

```
font-family = "MonaspiceRn Nerd Font Mono"
```

i have tried a variety of guides, commenting in and out, but any fallback for Chinese does not seem to use what is specified in the config

**Reply by Aaron-212** [reply]
Try this:
```
font-family = MonaspiceRn Nerd Font Mono
font-family = Lantinghei TC
```

**Reply by deeengan** [reply]
> Try this:
> 
> ```
> font-family = MonaspiceRn Nerd Font Mono
> font-family = Lantinghei TC
> ```

thanks. it is unfortunate, but that too does not seem to work for me. i can change many other things, like opacity etc. but it is the chinese font that seems to be stuck. i don't seem to have the bolding problem or instability that others do though. perhaps i will try again with your solution another way, another day 🌞

**Reply by deeengan** [reply]
> Try this:
> 
> ```
> font-family = MonaspiceRn Nerd Font Mono
> font-family = Lantinghei TC
> ```

ok i will say that after changing to tip it seems that i have the font i want working using your technique… fingers crossed, thanks

**Comment by shilohlee98** [comment]
This bug is somehow fixed in this release
https://ghostty.org/docs/install/release-notes/1-2-0#fallback-font-size-adjustment

**Reply by zeevenn** [reply]
still exists

**Comment by mitchellh** [comment, marked as answer]
This issue is very old, we've basically rewritten the font renderer since this, and I think there's too many non-actionable, non-specific issues in this one discussion. It's becoming a catch-all for all "this character doesn't look right" issues which isn't our goal here. We want to try to hone in on specific, actionable issues. As such, Im going to close this. If there are specific cases with the latest release, please reopen a new discussion

**Reply by yihuang** [reply]
It still exists on latest version (1.3.1) unfortunately.

**Reply by danielwe** [reply]
@yihuang see #9410 for the most informative and up-to-date tracking of this issue. This post seems to have made some progress toward a diagnosis: https://github.com/ghostty-org/ghostty/discussions/9410#discussioncomment-17038925.
